Block Support: Set border style none when border width zero

aaronrobertshaw opened the pull request:
~Depends on: https://github.com/WordPress/gutenberg/pull/31483~
Related issues: https://github.com/WordPress/gutenberg/issues/31333, https://github.com/WordPress/gutenberg/issues/31337
Related PRs: https://github.com/WordPress/gutenberg/pull/31585

## Description

As part of [refining the border support provided UI](https://github.com/WordPress/gutenberg/pull/31585), the border style dropdown is being replaced with a segmented control that does not include a `none` option. The idea is that setting a border width of `0` should apply the `border-style: none` property.

This PR updates the border width block support to update the border style attribute to `none` when the width is `0`. It then also re-applies the previous style selection if possible when the width is changed to a non-zero number, including being reset to undefined.

## How has this been tested?

#### Test Setup

Turn on custom borders within your theme.json file.
```json
			"border": {
				"customColor": true,
				"customRadius": true,
				"customStyle": true,
				"customWidth": true
			}
```

#### Test Instructions

1. Create a new post, add a group block to it and select the group block
2. Expand the border controls panel and select the solid border style option
3. Adjust the border width to a non-zero value, the solid border style button should remain on. 
4. Set the border width to `0`. The border style buttons should all be deselected.
5. Adjust the border width to a non-zero value again and the solid border style button should be reselected.
6. Update the border width to `0` again, switch to the code editor view and confirm that the border style attribute is now set to `none` and the appropriate style is applied within the markup.
2. Switch back to the visual editor, .
3. Change the border style selection to "Dashed" or "Dotted" and repeat the process adjusting the width value.
4. Confirm the block displays correctly on the frontend after saving.

## Screenshots 

https://user-images.githubusercontent.com/60436221/125031829-78d55b80-e0d0-11eb-9506-03f72a92a932.mp4


## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Set border style to none when 0 width

This application of the border-style: none will only occur if border style support has been opted into for the block.

ramonjd commented on packages/block-editor/src/hooks/border-width.js:
```diff
@@ -43,16 +54,34 @@ export const BorderWidthEdit = ( props ) => {
+		// border style prop to `none`. This style will only be applied if
+		// border style support has also been opted into.
+		if ( hasZeroWidth ) {
+			newStyle.border.style = 'none';
```
This is testing well 👍 

One thing I noticed: with the style removed, the inline styles on the frontend of my group block appear so:

`border-color:#f3ff98;border-radius:22px;border-style:none;border-width:0px;color:#1d1dff`

We're caching the previous border styles, so does it make sense to also remove `border-color` and associated class name `has-border-color`? 

I'm only thinking that, if the post is saved with zero width, the inline style and class for the border color remain unnecessarily forever.

aaronrobertshaw replied:
Thanks, I wasn't sure about how this might be used. I defaulted to changing as little as possible on the border block supports.

It could be possible to use the remaining border properties on child elements via inheriting them through CSS. I don't believe there is the ability to conditionally apply styles in block supports/global styles i.e. only apply the border color if border width > 0. That would mean I'll have to make a similar change as for style and unset the other border support options.

Perhaps I was overthinking things. I agree that it would be neater to remove the color, radius styles/classes etc if they aren't being used.

With the future refinements to the sidebar controls, in particular, the progressive discovery system, it might be that an entire border "instance" is added or removed. Which would effectively reset all border support properties in one go. I'm not sure at this point how that would play with theme set borders although I assume by the time that is implemented we'll have access to the theme set values for block contexts with the block editor, similar to what's available via the state in the full site editor.

In the end, this PR was to satisfy the [requested border support UI improvements](https://github.com/WordPress/gutenberg/issues/31337), and allow the user to override a theme set border and disable it. 

I'll update the PR to reset the other border properties. Thanks again for the thorough review 👍

aaronrobertshaw committed:
Clear border color selection on 0 border width

This clear border color selection when the border width is explicitly set to zero. When it is set back to a non-zero value (including undefined) it will attempt to restore the previous color selection if available.

aaronrobertshaw commented:
I've updated this to also clear the border color selection and restore it if available when switching from zero to non-zero border width. The same as occurs for the border style.

ramonjd commented:
> I've updated this to also clear the border color selection and restore it if available when switching from zero to non-zero border width. The same as occurs for the border style.

Nice! Retested and can confirm:

- the border color classes and inline styles (for custom border colors) are absent on the frontend when I set the border width to `0`
- when I add a new border width, my previously-selected border color is re-applied and the expected classes/inline styles appear on the frontend

ramonjd reviewed (approved):
LGTM

1. Border styles persist in the editor when adjusting width from `0` to `n` and back again (not between refreshes or switching between code and visual editors)
2. Frontend styles correspond to saved editor state.

-----------------------

Expand the border controls panel and select the solid border style option
Adjust the border width to a non-zero value, the solid border style button should remain on.

✅ 

Set the border width to 0. The border style buttons should all be deselected.

✅ 

Adjust the border width to a non-zero value again and the solid border style button should be reselected.

✅ 

Update the border width to 0 again, switch to the code editor view and confirm that the border style attribute is now set to none and the appropriate style is applied within the markup.

✅ 

Change the border style selection to "Dashed" or "Dotted" and repeat the process adjusting the width value.

✅ 

Confirm the block displays correctly on the frontend after saving.

✅ 


https://user-images.githubusercontent.com/6458278/125035109-bc31c900-e0d4-11eb-99c7-7e73f1f5b579.mp4

andrewserong reviewed (approved):
This is testing nicely for me @aaronrobertshaw!

✅ Setting the width to 0 updates the inline style in the markup to set `border-style: none;`
✅ While adjusting the width in the UI, my previous settings are remembered. I noticed that if I go to the code view to check my markup and then switch back, my previous border style _isn't_ remembered (since we're just using local React state rather than storing in the block's attributes). This seems fine to me — if I've switched out of editing the border, I don't expect it to remember my previous settings.
✅ The border style buttons render selected / vs deselected appropriately
✅ Works correctly with the dotted and dashed styles
✅ The block renders correctly on the front end

One other thing I noticed, which is separate to setting the border width to zero, is that if we have a non-zero value for border width, and de-select the border style, the `border-width` property still gets included in the inline style. Should that be omitted in that case?

![image](https://user-images.githubusercontent.com/14988353/125035746-89d49b80-e0d5-11eb-855b-0df9bbeb9e7c.png)

I think that's unrelated to this PR, so I don't think it's a blocker for moving this one forward.

aaronrobertshaw commented:
Thank you for the reviews @ramonjd  and @andrewserong.

> One other thing I noticed, which is separate to setting the border width to zero, is that if we have a non-zero value for border width, and de-select the border style, the border-width property still gets included in the inline style. Should that be omitted in that case?

With no border style selected it would be inheriting from the theme.json. In that case the adjustment of only the border width would still be valid. So in this instance I believe it should remain.

andrewserong commented:
> With no border style selected it would be inheriting from the theme.json. In that case the adjustment of only the border width would still be valid. So in this instance I believe it should remain.

Ah, great explanation. Thanks for the reminder! That sounds good to me, then. 👍
